### PR TITLE
wifi/iwd: handle IWD Station interface disappearing

### PR DESCRIPTION
**Pull request by balrog-kun**
The net.connman.iwd.Station interface, unlike the Device interface, can
go away and come back when the device changes modes or goes DOWN and UP.
The GDbusProxy for the interface becomes invalid after the interface
disappeared and reappeared.  This would make the IWD backend stop
working after rfkill was used or after suspend/resume (provided the
suspend/resume events are detected, without this everything works and is
really fast too).

Redo the handling of the Powered property changes, corresponding to
device UP state, to get a new GDBusProxy when the Station interface
reappears.  Simplify some checks knowing that priv->can_scan implies for
example that the Station interface is present, and that priv->enabled
implies the NM device state is >= DISCONNECTED.

### CODE REVIEW

**Comment by vhumpa** [comment]
NM CI: Please confirm test execution over this PR. Comment .*ok\W+to\W+test.* to confirm and/or .*add\W+to\W+whitelist.* to add the PR author in whitelist for future.

File: src/devices/wifi/nm-device-iwd.c [hunk 1]
```diff
@@ -1872,98 +1928,49 @@ nm_device_iwd_set_dbus_object (NMDeviceIwd *self, GDBusObject *object)
 	NMDeviceIwdPrivate *priv = NM_DEVICE_IWD_GET_PRIVATE (self);
 	GDBusInterface *interface;
 	GVariant *value;
+	gboolean powered;
 
 	if (!nm_g_object_ref_set ((GObject **) &priv->dbus_obj, (GObject *) object))
 		return;
 
+	if (priv->enabled && priv->dbus_station_proxy)
+		nm_device_queue_recheck_available (NM_DEVICE (self),
```
Comment (thom311) [inline src/devices/wifi/nm-device-iwd.c]: let's always use `{ }` when a block spans more than one line.

File: src/devices/wifi/nm-device-iwd.c [hunk 2]
```diff
@@ -1841,9 +1834,72 @@ station_properties_changed (GDBusProxy *proxy, GVariant *changed_properties,
...
+			/* No Station interface on the device object.  Check if the
+			 * "State" property is present on the Device interface, that
+			 * would mean we're dealing with an IWD version from before the
+			 * Device/Station split (0.7 or earlier) and we can easily
+			 * handle that by making priv->dbus_device_proxy and
+			 * priv->dbus_station_proxy both point at the Device interface.
+			 */
+			value = g_dbus_proxy_get_cached_property (priv->dbus_device_proxy, "State");
+			g_variant_unref (value);
+			if (!value) {
```
Comment (thom311) [inline src/devices/wifi/nm-device-iwd.c]: I think it's undefined behavior to check the value of dangling pointer, even in this case. At least, it's rather ugly.
Just move the unref after the "if" block.

**Comment by balrog-kun** [comment]
Ok, correcting both things.

**Comment by thom311** [comment]
merged to master.

btw, if you want any patches to be cherry-picked for nm-1-14 or nm-1-12 branches, please say so. That makes them eligible for the next minor releases (1.14.z or 1.12.z).

Thanks!

**Comment by Maryse47** [comment]
@thom311 as user who's facing an issue this patch fixes, I'm saying: please cherry-pick it for nm-1-14 branch.

**Comment by thom311** [comment]
Ok, now everything related to iwd is backported from master to nm-1-14 branch.

**Comment by Maryse47** [comment]
Great, thank you!

**Comment by balrog-kun** [comment]
Thank you, yes, it's going to be very useful to have these fixes in a minor release.

**Pull request closed without merging**